lib/handler/pool.ex

```diff
@@ -21,19 +35,27 @@ defmodule Handler.Pool do
     {:ok, state}
   end
 
-  def handle_call({:run, fun, opts}, from, state) do
-    case State.start_worker(state, fun, opts, from) do
-      {:ok, state} ->
-        {:noreply, state}
+  def handle_call({:run, fun, opts}, {pid, _tag}, state) do
+    case State.start_worker(state, fun, opts, pid) do
+      {:ok, state, ref} ->
+        {:reply, {:ok, ref}, state}
 
       {:reject, exception} ->
         {:reply, {:reject, exception}, state}
     end
   end
 
   def handle_info({ref, result}, state) when is_reference(ref) do
-    state = State.send_response(state, ref, result)
-    {:noreply, state}
+    if delegating_work?(state) do
```

Comment:
This is one are that feels a bit weird, I didn't want to send back the `{:DOWN, ...}` messages for when a task has actually died because that means we would need the clients to expect a 3rd response message, or it would mean that composed pools use a separate API than the main public API.

So for the sake of simplicity I'm having the shared pools release their resource commitments as soon they receive a response. The root pool will still wait for the process to actually die before releasing it's allocations, so you still have limits in place to protect you from over-allocating, but there will be a period of time where the composed pool has released its allocations and the root pool still has them earmarked.